tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/CodeownersTools/CodeownersTools.cs

```diff
@@ -0,0 +1,445 @@
+                    if (serviceLabelValidationResults != LabelHelper.ServiceLabelStatus.Exists)
+                    {
+                        var labelsPullRequests = await githubService.SearchPullRequestsByTitleAsync(Constants.AZURE_OWNER_PATH, Constants.AZURE_SDK_TOOLS_PATH, "Service Label");
+
+                        if (!LabelHelper.CheckServiceLabelInReview(labelsPullRequests, serviceLabel) && string.IsNullOrEmpty(normalizedPath))
+                        {
+                            throw new Exception($"Service label: {serviceLabel} doesn't exist.");
+                        }
+                    }
+                }
+
+                if (string.IsNullOrEmpty(workingBranch))
+                {
+                    var codeownersPullRequests = await githubService.SearchPullRequestsByTitleAsync(Constants.AZURE_OWNER_PATH, repo, "CODEOWNERS");
+
+                    foreach (var codeownersPullRequest in codeownersPullRequests)
+                    {
+                        if (codeownersPullRequest != null &&
+                            ((!string.IsNullOrEmpty(serviceLabel) && codeownersPullRequest.Title.Contains(serviceLabel, StringComparison.OrdinalIgnoreCase)) ||
+                            (!string.IsNullOrEmpty(normalizedPath) && codeownersPullRequest.Title.Contains(normalizedPath, StringComparison.OrdinalIgnoreCase))))
+                        {
+                            workingBranch = codeownersPullRequest.Head.Ref;
+                        }
+                    }
+                }
+
+                // Get codeowners file contents.
+                var codeownersFileContent = await githubService.GetContentsSingleAsync(Constants.AZURE_OWNER_PATH, repo, Constants.AZURE_CODEOWNERS_PATH);
+
+                if (codeownersFileContent == null)
+                {
+                    throw new Exception("Could not retrieve CODEOWNERS file from the repository.");
+                }
+
+                var branchToFetch = string.IsNullOrEmpty(workingBranch) ? "main" : workingBranch;
+                var codeownersUrl = $"{githubRawContentBaseUrl}/{Constants.AZURE_OWNER_PATH}/{repo}/{branchToFetch}/{Constants.AZURE_CODEOWNERS_PATH}";
+
+                var codeownersContent = codeownersFileContent.Content;
+                var codeownersSha = codeownersFileContent.Sha;
+
+                // Find Codeowner Entry with the validated Label or Path
+                bool isMgmtPlane = typespecHelper.IsTypeSpecProjectForMgmtPlane(typeSpecProjectRoot);
+                var (startLine, endLine) = isMgmtPlane
+                    ? codeownersHelper.FindBlock(codeownersContent, standardManagementCategory)
+                    : codeownersHelper.FindBlock(codeownersContent, standardServiceCategory);
+
+                var codeownersEntries = CodeownersParser.ParseCodeownersFile(codeownersUrl, azureWriteTeamsBlobUrl, startLine, endLine);
```

Comment:
`codeownersUrl` could be different from the content fetched in `codeownersFileContent`. One is always from `main` the other might be from the state of some PR.